Consolidated the two test exes into one

SteveL-MSFT opened the pull request:
In preparation for another PR where I need another test exe, it made sense to have just a single one rather than multiple.  Combined the previous echoargs and createchildprocess test exes into a single TestExe that can be extended for other tests.

SteveL-MSFT committed:
Consolidated the two test exes into one

iSazonov commented on test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:
```diff
@@ -1,41 +1,41 @@
+    while ($powershellTestDir -notmatch 'test[\\/]powershell$') {
+        $powershellTestDir = Split-Path $powershellTestDir
+    }
+    $testexe = Join-Path (Split-Path $powershellTestDir) tools/TestExe/bin/testexe
```
I see the "find testsexe" code is used multiple times in tests. Maybe move it in HelpersCommon.psm1 as a function?
And we should use BeforeAll here.

SteveL-MSFT replied:
Good suggestion.  I was only modifying the existing code for the new exe without changing any existing logic.  Similar to how we add the test modules to PSModulePath, I'm thinking of just adding the testexe to PATH.

iSazonov replied:
Good catch!

Closed.

iSazonov commented on build.psm1:
```diff
@@ -681,13 +681,13 @@ function Publish-PSTestTools {
     Find-Dotnet
 
-    $tools = @("$PSScriptRoot/test/tools/EchoArgs", "echoargs"), @("$PSScriptRoot/test/tools/CreateChildProcess", "createchildprocess")
+    $tools = ,@("${PSScriptRoot}/test/tools/TestExe", "testexe")
```
TestExe vs testexe - We need to be consistent and cautious with the case of letters.
TestExe may be a bad name on Linux. Maybe `BinaryTestRunner` or something like?

SteveL-MSFT replied:
I was trying to follow the existing convention for EchoArgs and CreateChildProcess.  Since DotNet calls it an exe, I think TestExe is appropriate.  "TestRunner" to me means something that runs tests.

iSazonov replied:
Closed.

iSazonov commented on test/powershell/Language/Scripting/NativeExecution/NativeCommandProcessor.Tests.ps1:
```diff
@@ -42,20 +42,20 @@ Describe "Native Command Processor" -tags "Feature" {
         # make sure no test processes are running
-        # on Linux, the Process class truncates the name so filter using Where-Object
-        Get-Process | Where-Object {$_.Name -like 'createchildproc*'} | Stop-Process
+        Get-Process testexe -ErrorAction SilentlyContinue | Stop-Process
```
Docs say that by default, Stop-Process prompts for confirmation - we should add "-Force".

SteveL-MSFT replied:
Docs must be wrong, it doesn't prompt by default unless you specify `-confirm`.  I'll submit a doc fix.

SteveL-MSFT replied:
https://github.com/PowerShell/PowerShell-Docs/pull/1276

iSazonov replied:
Closed.

iSazonov commented on test/powershell/Language/Scripting/NativeExecution/NativeCommandProcessor.Tests.ps1:
```diff
@@ -72,7 +72,7 @@ Describe "Native Command Processor" -tags "Feature" {
                 break
```
It seems we can use Wait-Process -Timeout instead of "while" cycle.

SteveL-MSFT replied:
I didn't make changes to the test. I think if you have concerns about the test, we should handle that as a separate issue and PR. I'm only trying to converge to a single test executable.

iSazonov replied:
I left a comment in #2561.

Closed.

iSazonov commented on test/powershell/Language/Scripting/NativeExecution/NativeCommandProcessor.Tests.ps1:
```diff
@@ -72,7 +72,7 @@ Describe "Native Command Processor" -tags "Feature" {
-        $childprocesses = Get-Process | Where-Object {$_.Name -like 'createchildproc*'}
+        $childprocesses = Get-Process testexe
         $count = $childprocesses.count
         $childprocesses | Stop-Process
```
Again "Stop-Process -Force"

SteveL-MSFT replied:
Doc is wrong

iSazonov replied:
Closed.

iSazonov commented on test/powershell/Language/Scripting/NativeExecution/NativeStreams.Tests.ps1:
```diff
@@ -1,70 +1,70 @@
-Describe "Native streams behavior with PowerShell" -Tags 'CI' {
```
Why was the Describe removed and re-added?

SteveL-MSFT replied:
Not sure why the diff is showing up this way.  I made edits directly replacing EchoArgs in this case.

iSazonov replied:
Closed.

iSazonov commented on test/powershell/Language/Scripting/NativeExecution/NativeStreams.Tests.ps1:
```diff
@@ -1,70 +1,70 @@
+            Start-Sleep -Milliseconds 100
+            # yeild some multi-line formatted object
+            @{'a' = 'b'}
+        } | & $testexe -echoargs | Should Be $null
```
I don't understand the test. It seems it does not match its description.

-

SteveL-MSFT replied:
I didn't make changes to the test.  I think if you have concerns about the test, we should handle that as a separate issue and PR.  I'm only trying to converge to a single test executable.

SteveL-MSFT replied:
I think it's because VSCode change the newline from LF to CRLF

iSazonov replied:
Closed.

iSazonov commented on test/tools/TestExe/TestExe.cs:
```diff
@@ -0,0 +1,47 @@
+                switch(args[0])
+                {
+                    case "-echoargs":
+                        for (int i = 1; i < args.Length; i++)
```
Maybe put the test code in each case in a function if we plan to add multiple tests?
And it would be good to add comments describing the purpose of the test code.

SteveL-MSFT replied:
I'm fine with that, will change.

iSazonov replied:
Closed.

iSazonov reviewed (commented):
Leave a comment

iSazonov commented:
@SteveL-MSFT Sorry, most of my comments go beyond your changes.

SteveL-MSFT committed:
updated build.psm1 to add testexe to path on publish of test tools
moved test specific code to separate methods in testexe
removed need to discover location of testexe

SteveL-MSFT commented:
@lzybkr can you merge?  thanks

lzybkr commented on test/tools/TestExe/TestExe.cs:
```diff
@@ -0,0 +1,64 @@
+            if (args.Length > 0)
+            {
+                switch(args[0])
+                {
```
Maybe add `.ToLowerInvariant` to save folks some hassle if they use mixed case.

SteveL-MSFT replied:
sounds good

lzybkr commented on build.psm1:
```diff
@@ -681,25 +681,29 @@ function Publish-PSTestTools {
 
-    $tools = @("$PSScriptRoot/test/tools/EchoArgs", "echoargs"), @("$PSScriptRoot/test/tools/CreateChildProcess", "createchildprocess")
+    $tools = ,@("${PSScriptRoot}/test/tools/TestExe", "testexe")
     if ($Options -eq $null)
```
This shouldn't be an array of arrays, it should be an array of custom objects.

SteveL-MSFT replied:
How about an array of hashtables?

lzybkr replied:
That works too, but `[pscustomobject]@{Path="${PSScriptRoot}/test/tools/TestExe";Output="testexe"}` expresses the intent more clearly to me.

---
In reply to: [121528687](https://github.com/PowerShell/PowerShell/pull/3982#discussion_r121528687) [](ancestors = 121528687)

SteveL-MSFT committed:
Changed tools list from array of arrays to using array of hashtable
Do arg check on testexe as lowercase
